### PR TITLE
FluidDynamics - added 3d newtonian law with nodal properties

### PR DESCRIPTION
**Pull request by ddiezrod**
This is a proposal to have a newtonian law that can be used for a two fluid element where the properties (dynamic viscosity in this case) are read from the nodes.
It also includes the smagorinsky term in order to compute the effective viscosity.

Thoughts? @jrubiogonzalez @jcotela @rubenzorrilla @RiccardoRossi @pooyan-dadvand

### CODE REVIEW

File: applications/FluidDynamicsApplication/custom_constitutive/newtonian_3d_law.cpp [hunk 1]
```diff
@@ -106,6 +104,11 @@ double Newtonian3DLaw::GetEffectiveViscosity(ConstitutiveLaw::Parameters& rParam
     return rParameters.GetConstitutiveMatrix()(5,5);
 }
 
+double Newtonian3DLaw::ComputeEffectiveViscosity(ConstitutiveLaw::Parameters& rParameters) const {
+    const double effective_viscosity = rParameters.GetMaterialProperties()[DYNAMIC_VISCOSITY];
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_constitutive/newtonian_3d_law.cpp]: You can avoid declaring the variable by doing
```cpp
return rParameters.GetMaterialProperties()[DYNAMIC_VISCOSITY];
```

File: applications/FluidDynamicsApplication/custom_constitutive/newtonian_3d_law.cpp [hunk 2]
```diff
@@ -106,6 +104,11 @@ double Newtonian3DLaw::GetEffectiveViscosity(ConstitutiveLaw::Parameters& rParam
     return rParameters.GetConstitutiveMatrix()(5,5);
 }
 
+double Newtonian3DLaw::ComputeEffectiveViscosity(ConstitutiveLaw::Parameters& rParameters) const {
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_constitutive/newtonian_3d_law.cpp:107]: I think that virtual specifier is missing...

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
I'd do the implementation of the new constitutive law in a .cpp file. Add a simple .cpp test (have a look to the existen ones) and I'll be ok with the changes.

Thanks in advance!

File: applications/FluidDynamicsApplication/custom_constitutive/newtonian_3d_law.cpp (hunk 2 above)
Comment (ddiezrod) [reply applications/FluidDynamicsApplication/custom_constitutive/newtonian_3d_law.cpp:107]: it is in the header. I think it should only be there, right?

File: applications/FluidDynamicsApplication/custom_constitutive/newtonian_3d_law.cpp (hunk 1 above)
Comment (ddiezrod) [reply applications/FluidDynamicsApplication/custom_constitutive/newtonian_3d_law.cpp]: ok, done. Thanks!

File: applications/FluidDynamicsApplication/custom_constitutive/newtonian_nodal_properties_3d_law.h [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+        if (prop.Has(C_SMAGORINSKY)) {
+            const double csmag = prop[C_SMAGORINSKY];
+            if (csmag > 0.0) {
+                const double strain_rate = EquivalentStrainRate(rParameters);
+                const BoundedMatrix<double, 4, 3>& DN_DX = rParameters.GetShapeFunctionsDerivatives();
+                const double elem_size = ElementSizeCalculator<3,4>::GradientsElementSize(DN_DX);
+                double length_scale = csmag * elem_size;
+                length_scale *= length_scale;
+                viscosity += 2.0*length_scale * strain_rate;
```
Comment (jcotela) [inline applications/FluidDynamicsApplication/custom_constitutive/newtonian_nodal_properties_3d_law.h]: Be careful here, the Smagorinsky model gives you a kinematic viscosity. If `C_SMAGORINSKY` is a dimensionless number, you need to multiply times the density if you want to add it to a dynamic viscosity.

**Review by jcotela — CHANGES_REQUESTED** [review]
Please check the comments, I think there is a unit problem. Other than that, I'm OK with the additions, but I'm not convinced about the name of the new class, I'd prefer something that makes it explicit that the law also assumes the existence of a level-set distance. Maybe Newtonian3DDiscontinuous or Newtonian3DTwoFluid?

**Comment by ddiezrod** [comment]
@jcotela I think you are probably right about the name. The idea at the beginning was to make a constitutive law that used the nodal properties, but then I realized that more changes were necessary and at the end it became something else. If it is ok for everyone I will rename it NewtonianTwoFluid3DLaw.

File: applications/FluidDynamicsApplication/custom_constitutive/newtonian_nodal_properties_3d_law.h (hunk 3 above)
Comment (ddiezrod) [reply applications/FluidDynamicsApplication/custom_constitutive/newtonian_nodal_properties_3d_law.h]: You are right! Fixed

**Review by jcotela — DISMISSED** [review]
:+1: Thanks for the fixes.

**Comment by ddiezrod** [comment]
@rubenzorrilla cpp test added !

**Comment by rubenzorrilla** [comment]
Please, do the implementation in a .cpp (just to follow the other CLaws standards). Thanks in advance!

**Comment by ddiezrod** [comment]
@rubenzorrilla done. Thanks!

**Review by rubenzorrilla — APPROVED** [review]

**Pull request merged**